### DISCUSSION TITLE
How to set sample_rate in i2s_config_t to not be updated automatically by A2DP

### DISCUSSION THREAD

**Discussion by abarrettFrontier** [pschatzmann/ESP32-A2DP#602] [Q&A] [answered]
I am working with an esp32 pico mini to collect Bluetooth from a phone and pass that data via I2S to another device. I am setting the sample_rate to 48000 in the i2s_config_t, however, once my phone connects, the rate drops down to 44100. Our second device is not compatible with 44100, so I need a way to ensure that it will continue to transmit on the correct rate. This is how my code looks currently:

` Serial.begin(115200);
  Wire.begin(I2C_SDA, I2C_SCL);
  a2dp_sink.set_auto_reconnect(true);
 
  //Set I2S parameters
  static i2s_config_t i2s_config = {
    .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX),
    .sample_rate = 48000, // updated automatically by A2DP
    .bits_per_sample = (i2s_bits_per_sample_t)32,
    .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
    .communication_format = (i2s_comm_format_t) (I2S_COMM_FORMAT_STAND_I2S),
    .intr_alloc_flags = 0, // default interrupt priority
    .dma_buf_count = 8,
    .dma_buf_len = 64,
    .use_apll = true,
    .tx_desc_auto_clear = true // avoiding noise in case of data unavailability
  };

  i2s_pin_config_t I2S_devKit_pin_config = {
    .bck_io_num = 12,
    .ws_io_num = 13,
    .data_out_num = 15,
    .data_in_num = I2S_PIN_NO_CHANGE
  };

  a2dp_sink.set_pin_config(I2S_devKit_pin_config);
  a2dp_sink.set_i2s_config(i2s_config);
  a2dp_sink.start("Esp Device");`

**Comment by pschatzmann** [comment, marked as answer]
The sample rate is defined by the ESP32 A2DP and can't be changed.

I don't understand what you mean with second device, but if your output DAC only supports 48000, you need to resample. The AudioTools library provides the functionality and documentation how to do this!

**Reply by abarrettFrontier** [reply]
I believe the sample rate is changing. When the bluetooth dev kit was first powered on I used an osiliscope to check the sample rate and it continuously read at 48000 until I connected a device such as a phone or laptop at which point it instead transmitted at 44100. Do you mean it cannot be changed after that point?

By second device I mean: we have the esp32 dev kit connected to our custom microcontroller via I2S and I am trying to transmit the bluetooth data that the esp32 receives over to our microcontroller to be handled before outputting it. That transmission needs to happen over I2S, however the microcontroller cannot use a sampling rate of 44100. 

I will look into the resampling. Alternatively, could I set the esp32 to be a Slave instead of Master and transmit the data while allowing the microcontroller to set the rate? If so, are there any examples that you could highlight that follow this path?

**Comment by pschatzmann** [comment]
You need to use the sample rate that the data is provided by A2DP! 
You can query the rate after via the API after the A2DP connection has been established but I have never seen anything else then 44100.

Did you test if you get a good I2S connection between 2 different microcontrollers ? 
I never could make this work reliably...

**Reply by abarrettFrontier** [reply]
2 different microcontrollers meaning the esp devkit and my second microcontroller? I have not yet had success, I received the data but it is coming out garbled which I had assumed was due to the missmatch of data rates. 
I have since noticed that both ends of the word select (sample rate) channel were outputting different rates (44100 from the esp and 48000 from the 2nd microcontroller), when it should only be the one so I decided to try setting the esp32 to slave, but now no audio is being produced at all. Which is why I was wondering if you had any examples that did this, but it sounds like you do not.

**Comment by pschatzmann** [comment]
You can use the [AudioTools to test this independently](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working) of the A2DP

**Comment by abarrettFrontier** [comment]
I am having difficulty getting the resample working and outputting to the I2S stream, would you be able to provide an example that does this? Or even just how to set the input to be the audio data received via Bluetooth.

**Comment by pschatzmann** [comment]
With the actual version it is pretty easy: just send the output to a ResampleStream instead of an I2SStream and don't forget to start all objects:

```C++
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AudioInfo from_info(44100, 2, 16);
AudioInfo to_info(48000, 2, 16);
I2SStream out;
ResampleStream resample(out);
BluetoothA2DPSink a2dp_sink(resample);

void setup() {
  // setup i2s with 48000
  auto cfg = out.defaultConfig();
  cfg.copyFrom(to_info);
  cfg.pin_bck = 12;
  cfg.pin_ws = 13;
  cfg.pin_data = 15;
  cfg.is_master = true; // or false for slave
  out.begin(cfg);

  // setup resample from 42100 to 48000
  resample.begin(from_info, to_info.sample_rate);
  
  // start a2dp
  a2dp_sink.start("Esp Device");  
}


void loop() {
  delay(1000); // do nothing
}
```

**Reply by abarrettFrontier** [reply]
Okay thank you, this makes a lot of sense, I look forward to testing it with my device tomorrow! The only thing I still have a question about is how am I able to differenciate to the module if this is the Master/Slave mode on the i2s stream? I need the esp device to be in i2s slave mode {ex: .mode = (i2s_mode_t) (I2S_MODE_SLAVE | I2S_MODE_TX).}

**Reply by pschatzmann** [reply]
I updated the example above

**Comment by abarrettFrontier** [comment]
So, the performance has improved, I can now hear the music playing, it is just accompanied by lots of static and some volume changes. I believe part of my issue is because on the other end I have a differing slot width and word size (slot_width = 32 and word_size = 24) is there a way to accommodate that in your code while formatting the data to align to the left?

**Reply by pschatzmann** [reply]
A2DP only supports 16 bits.
If you finally output to I2S you can define the [i2s_format](https://pschatzmann.github.io/arduino-audio-tools/namespaceaudio__tools.html#a720616ce211566d808f3e441f9b18cda)
